### ISSUE TITLE
Crashpad linking error on x86_64 windows

### ISSUE THREAD

**Issue by petrmanek** [getsentry/sentry-native#939]
**Description**

Since updating to 0.7.0, I have been getting a crashpad linking error (see log for details). It appears to be caused by [this symbol](https://github.com/getsentry/crashpad/blob/3182e3be21a8a753f9f269f0a590370d49c8f3cf/util/misc/capture_context_win.asm#L219) not being declared probably due to [this macro](https://github.com/getsentry/crashpad/blob/3182e3be21a8a753f9f269f0a590370d49c8f3cf/util/misc/capture_context_win.asm#L17) not being defined. I am uncertain what causes that. The macro in question is supposed to detect the x64 architecture, which I am targeting. Specifically, I invoke cmake with `-A x64` and vcvarsall.bat with `amd64` prior to build.

Alternatively, this could be caused by my project building the referenced assembly file, and later failing to include it in linked objects. But I am uncertain why would that happen as it is reference [from here](https://github.com/getsentry/crashpad/blob/3182e3be21a8a753f9f269f0a590370d49c8f3cf/util/CMakeLists.txt#L394). Interestingly, though, I cannot find any mentions of the object file `capture_context_win.asm` in my build logs prior to failure, so that could be perhaps a smoking gun.

**When does the problem happen**

- [x] During build
- [ ] During run-time
- [ ] When capturing a hard crash

**Environment**

- OS: Windows 10 (21H2), 64-bit (x86_64), SDK 10.0.22621.0
- Compiler: MSVC 19.38.33134.0
- CMake version and config: 3.26, `SENTRY_ENABLE_INSTALL=ON`, `SENTRY_BUILD_EXAMPLES=ON`

**Steps To Reproduce**

1. Include sentry in a cmake project via `add_subdirectory()`, and make an arbitrary executable target depend on it.
2. Build on the specified platform with the specified toolchain.
3. Observe error during build.

**Log output**

```
crashpad_snapshot.lib(process_reader_win.obj) : error LNK2019: unresolved external symbol "void __cdecl crashpad::CaptureContext(struct _CONTEXT *)" (?CaptureContext@crashpad@@YAXPEAU_CONTEXT@@@Z) referenced in function "bool __cdecl crashpad::`anonymous namespace'::FillThreadContextAndSuspendCount<struct crashpad::process_types::internal::Traits64>(void *,void *,struct crashpad::ProcessReaderWin::Thread *,enum crashpad::ProcessSuspensionState,bool)" (??$FillThreadContextAndSuspendCount@UTraits64@internal@process_types@crashpad@@@?A0x57c906c6@crashpad@@YA_NPEAX0PEAUThread@ProcessReaderWin@1@W4ProcessSuspensionState@1@_N@Z) [C:\GitLab-Runner\builds\SvfzDLzP\0\app\build\lib\sentry\crashpad_build\handler\crashpad_handler.vcxproj]
crashpad_client.lib(crashpad_client_win.obj) : error LNK2001: unresolved external symbol "void __cdecl crashpad::CaptureContext(struct _CONTEXT *)" (?CaptureContext@crashpad@@YAXPEAU_CONTEXT@@@Z) [C:\GitLab-Runner\builds\SvfzDLzP\0\app\build\lib\sentry\crashpad_build\handler\crashpad_handler.vcxproj]
C:\GitLab-Runner\builds\SvfzDLzP\0\app\build\RelWithDebInfo\crashpad_handler.exe : fatal error LNK1120: 1 unresolved externals [C:\GitLab-Runner\builds\SvfzDLzP\0\app\build\lib\sentry\crashpad_build\handler\crashpad_handler.vcxproj]
```

**Comment by petrmanek** [comment]
Found [a candidate solution](https://stackoverflow.com/a/38794302). It is possible that without `/safeseh` MSASM compiler just silently ignores all .asm sources (yay, Windows). Testing it now, will report back.

**Comment by petrmanek** [comment]
So, as suggested by the SO answer, I tried adding the following snippet to `external/crashpad/util/CMakeLists.txt`:
```cmake
        set_source_files_properties(
            misc/capture_context_win.asm
            win/safe_terminate_process.asm
            PROPERTIES
            COMPILE_FLAGS "/safeseh"
        )
```

It did not help. I verified that the statement is being executed using a debug message, but it does not seem to have any effect. It is now clear the linker is correctly complaining about missing symbols because the referenced MSASM files are never compiled. Could this be a cmake bug?

**Comment by petrmanek** [comment]
Having read a couple more discussions on this topic, I discovered [this thread](https://discourse.cmake.org/t/building-lib-file-from-asm-cmake-bug/1959). I can confirm that the following patch to `external/crashpad/util/CMakeLists.txt` fixes the problem for me:

```diff
--- sentry/external/crashpad/util/CMakeLists.txt	2024-01-19 11:44:18.699933982 +0100
+++ bugfix_crashpad_CMakeLists.txt	2024-01-19 14:29:08.965482626 +0100
@@ -389,11 +389,22 @@
         target_sources(crashpad_util PRIVATE
             misc/capture_context_win_arm64.asm
         )
+        set_source_files_properties(
+            misc/capture_context_win_arm64.asm
+            PROPERTIES
+            LANGUAGE ASM_MASM
+        )
     else()
         target_sources(crashpad_util PRIVATE
             misc/capture_context_win.asm
             win/safe_terminate_process.asm
         )
+        set_source_files_properties(
+            misc/capture_context_win.asm
+            win/safe_terminate_process.asm
+            PROPERTIES
+            LANGUAGE ASM_MASM
+        )
     endif()
 endif()
```

I may have accused Microsoft toolchain falsely after all! This seems to be a cmake bug where the .asm file extension is not automatically mapped to `LANGUAGE ASM_MASM`, so the build system simply skips these files and does not emit any warnings or error messages. Super weird.

In any case, I have my workaround for now. I wanted to originally implement this as a patch, but since my plain Windows environment does not have anything like `diff` or `patch`, I am overwriting the referenced file with my locally patched version using `file(COPY_FILE ...)` prior to including Sentry as a subproject in my `CMakeLists.txt`.

Please feel free to file a bug upstream.

**Comment by petrmanek** [comment]
So I have managed to test my proposed solution with arm64 as well. This entailed a surprising detour which led me to discover that when one wants to compile `ASM` on Windows, one has to enable a specific cmake language depending on the target architecture; either `ASM_MASM` or `ASM_MARMASM`. Contrast that with just enabling `ASM` on any sane Linux platform.

When this is done incorrectly (i.e. `ASM_MASM` is enabled on arm64 or `ASM_MARMASM` is enabled on amd64), there is a silent but ominous message during configuration, which says that the found assembly compiler is `unknown` rather than `MSVC`. This does not produce a configuration error, but much later the build fails with the following message:

```
error MSB3721: The command "echo MASM not supported on this platform
```

One would think that the best way to select the `ASM` flavor depending on target architecture would be to check the value of `${CMAKE_SYSTEM_PROCESSOR}` but that is yet another fallacy. As it turns out, on Windows the value of this variable is effectively fixed to `AMD64` even when targeting arm64. This is a [known cmake bug](https://gitlab.kitware.com/cmake/cmake/-/issues/15170) that has yet to be resolved. The referenced ticket states that for the time being users are advised not ti rely on the broken cmake variable and instead use other variables to correctly configure projects, some of which are internal or undocumented and may disappear in future cmake versions.

In the end, I decided to use the `VSCMD_ARG_TGT_ARCH` environment (not cmake) variable to make my determination. Since I use `vcvarsall.bat`, this variable can be relied upon to be declared and correctly point to the target architecture. Updating my patch to reflect these learnings, we get something like:

```diff
--- sentry/external/crashpad/util/CMakeLists.txt	2024-01-19 11:44:18.699933982 +0100
+++ bugfix_crashpad_util_CMakeLists.txt	2024-01-20 17:25:40.759199970 +0100
@@ -385,15 +385,26 @@
         win/traits.h
         win/xp_compat.h
     )
-    if("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "arm|ARM64")
+    if("$ENV{VSCMD_ARG_TGT_ARCH}" MATCHES "arm|ARM64")
         target_sources(crashpad_util PRIVATE
             misc/capture_context_win_arm64.asm
         )
+        set_source_files_properties(
+            misc/capture_context_win_arm64.asm
+            PROPERTIES
+            LANGUAGE ASM_MARMASM
+        )
     else()
         target_sources(crashpad_util PRIVATE
             misc/capture_context_win.asm
             win/safe_terminate_process.asm
         )
+        set_source_files_properties(
+            misc/capture_context_win.asm
+            win/safe_terminate_process.asm
+            PROPERTIES
+            LANGUAGE ASM_MASM
+        )
     endif()
 endif()
```

Furthermore, a project-wide `ASM` flavor selection occurs for sentry one level above, so in addition to `external/crashpad/util/CMakeLists.txt` I also needed to patch `external/crashpad/CMakeLists.txt` as follows:

```diff
--- sentry/external/crashpad/CMakeLists.txt	2024-01-20 17:26:51.662726063 +0100
+++ bugfix_crashpad_CMakeLists.txt	2024-01-20 17:44:58.865995394 +0100
@@ -50,7 +50,7 @@
 endfunction()

 if(WIN32)
-    if("${CMAKE_SYSTEM_PROCESSOR}" MATCHES ARM64)
+    if("$ENV{VSCMD_ARG_TGT_ARCH}" MATCHES "arm|ARM64")
         enable_language(ASM_MARMASM)
     else()
         enable_language(ASM_MASM)
```

With both of these patches in place, everything started to compile again for both CPU architectures! :tada: 

**One final note:** I would imagine that there are other components of sentry-native that rely on `ASM` on Windows. A quick `grep` reveals that similar locations are present in breakpad, and indeed [in sentry-native too](https://github.com/getsentry/sentry-native/blob/9ea60904642288abf6e90a4c7d37e142af399953/CMakeLists.txt#L19). I already spent way too much time going down this particular rabbit hole, but you may want to recursively `grep` all your `*.txt` files for occurrences of `ASM` and introduce similar logic.

Hope this helps!

**Comment by supervacuus** [comment]
Thanks, @petrmanek, for the report. I would like to understand better how your setup leads to your issues. We're currently building the Native SDK in various configurations and never ended up with ignored assembly sources in our `MSVC` builds. Do you have a reference for a `CMake` bug that would ignore global language settings for some source file types? Is it possible that you're starting the `CMake` configuration with incomplete toolchain information? Can you provide your CMake (config+build) invocation and output for comparison?

While we currently do not officially support `ARM64` on Windows (because we have no stable test-runners for that platform), [we already build](https://github.com/getsentry/crashpad/actions/runs/7115516889/job/19371855513) `crashpad` for `ARM64` on Windows to ensure that we stay intact when we modify our fork or update from upstream. If the assembler is not correctly identified during configuration, the `ASM` language settings will be ignored for the whole project.

We know the CMake issue around `CMAKE_SYSTEM_PROCESSOR` when using the MSVC generator. The right approach for this is to provide a [toolchain configuration](https://github.com/getsentry/crashpad/blob/89991e9910bc4c0893e45c8cfad0bdd31cc25a5c/cmake/toolchains/win_arm64.cmake) instead of relying on an environment variable of the generator-target because the build script should express those pesky differences as much as possible in CMake's meta-description. 

Be aware that you must provide the target twice (once for the _configuration_ and once for the _build_ step) when using the MSVC generator and providing`CMAKE_GENERATOR_PLATFORM` rather than `-A` on the command line:

https://github.com/getsentry/crashpad/blob/getsentry/.github/workflows/build.yml#L37-L47

Just to let you know, with this approach, you can use the same toolchain file when building with MSVC and MinGW, which should be orthogonal to targeting ARM64 on Windows.

**Comment by petrmanek** [comment]
@supervacuus I will try to address your questions in order:

> Do you have a reference for a CMake bug that would ignore global language settings for some source file types?

I am not sure how this could happen but it seems to me that (at least in my toolchain, which is nothing out of the ordinary) the `.asm` file extension is not mapped like you expect. I could not find a CMake issue about this per se, but for instance [this discussion thread](https://discourse.cmake.org/t/compiling-arm64-asm-in-visual-studio-with-cmake/500) appears to describe exactly the same problem I am having. Furthermore, if you look up keywords like "cmake msvc masm linking error", you will find at least 10 similar threads. Interestingly, about half of them is solved by adding `/safeseh`, which I tried and did not work.


> Is it possible that you're starting the CMake configuration with incomplete toolchain information?

I am uncertain how we would qualify what it means to have _complete_ information. I come from a POSIX land, so it is possible that I forgot something, but in my build batch script the following happens for `x64` (and for `arm64` in parentheses):

1. The `vcvarsall.bat` script is called with `amd64` (or `am64_arm64`) to set up the developer command prompt.
2. VCPKG configuration is selected: `VCPKG_HOST_TRIPLET=x64-windows` and `VCPKG_TARGET_ARCHITECTURE=x64` (or `VCPKG_TARGET_ARCHITECTURE=arm64`)
3. CMake architecture flag is selected: `BUILD_ARCH_FLAG="-A x64"` (or `BUILD_ARCH_FLAG="-A arm64"`)
4. CMake build system is configured with: `cmake %BUILD_ARCH_FLAG% -D CMAKE_BUILD_TYPE=RelWithDebInfo -D CMAKE_TOOLCHAIN_FILE=%CI_VCPKG_TOOLCHAIN% -D VCPKG_HOST_TRIPLET=%BUILD_VCPKG_HOST_TRIPLET% -D VCPKG_TARGET_ARCHITECTURE=%BUILD_VCPKG_ARCH% -S ..`
5. CMake project is built with: `cmake --build . --config RelWithDebInfo -- /p:CL_MPCount=8` (note that the `/p` flag is a workaround for [this issue](https://gitlab.kitware.com/cmake/cmake/-/issues/20564))
6. CMake project is installed into the artifact directory: `cmake --install . --config RelWithDebInfo --prefix artifacts`

As you can see, there are two environment-level setups here: one of them is the VS Developer Command Prompt, and the other is VCPKG's toolchain file, which in my understanding should respond to the defined CMake variables. If I drop my workarounds, the issue happens at stage 5 (linking at the end of project build). All these command are written defensively, so that they will halt on error (e.g. instead of `command` I write `command || goto :error`), so I am tempted to say that in this sense my toolchain information is complete. However, it is of course possible that I missed something stupid. Building things for distribution on Windows is still dark magic to me!


> Can you provide your CMake (config+build) invocation and output for comparison?

So I already shared my CMake invocation for the previous questions. Here I am attaching the configuration output for two x86_64 cases. One, where my bugfix patches are applied, and another where a vanilla version of sentry-native 0.7.0 is included in my project without any modifications. Short of some extra debug messages that I added in my patched version, those outputs are identical for `x86_64`. Curiously, if you look at the arm64-equivalent of the vanilla configuration, you can see that ASM compiler was detected as unknown, and only when I include sentry-native, it is detected as `ASM_MASM`, which is the incorrect version for `arm64`. Also note that short of the patched version, both vanilla versions skip assembling the referenced .asm files, and later complain during build.

[patched_config.log](https://github.com/getsentry/sentry-native/files/14011628/patched_config.log)
[vanilla_config.log](https://github.com/getsentry/sentry-native/files/14011629/vanilla_config.log)
[vanilla_config_arm64.log](https://github.com/getsentry/sentry-native/files/14011677/vanilla_config_arm64.log)


> The right approach for this is to provide a toolchain configuration instead of relying on an environment variable of the generator-target because the build script should express those pesky differences as much as possible in CMake's meta-description.

I am aware of this, and am mildly disappointed that VCPKG does not handle that when configured with `VCPKG_TARGET_ARCHITECTURE=arm64`. Not sure if I should be filing a bug upstream, what do you think?


> Be aware that you must provide the target twice (once for the configuration and once for the build step) when using the MSVC generator and providing CMAKE_GENERATOR_PLATFORM rather than -A on the command line:

So here I may be using incorrect invocation of CMake. As you see above, I use only `-A` at configuration stage along with CMake variables, which are directed to VCPKG. The only parameter which appears during configuration _as well as_ build, is the build configuration `RelWithDebInfo`. One question though, in the referenced section of your YAML file, I can see no mention of `CMAKE_GENERATOR_PLATFORM`. Instead I see `/p:Platform=ARM64`. Are these equivalent?

Hope this helps!

**Comment by supervacuus** [comment]
Thanks @petrmanek!

I had considerable issues with `vcpkg` and toolchain configurations that deviate from the norm, up to the point where I had to fully uninstall `vcpkg` to get `clang-cl` to work in a CMake project using the MSVC generator (in a project that didn't even use `vcpkg` directly, just having it installed led to issues). 

Right now, we do not have the capacity to test against yet another layer of toolchain configuration, given that we already support a significant range of toolchains, platforms, and architectures. So, while I would be more than willing to accept PRs that fix issues in that context, these would still have to be in a way that doesn't depend on environmental configurations where we dive through multiple layers of abstractions (i.e., access variables that will exist if a specific generator configures in the context of two other environment layers).

The good news is that I am almost certain that your build should work without modifications to the build script if you can make your toolchains discoverable for CMake (we build for AMD64 and ARM64 in CI and know of these issues only from toolchain misconfigurations).

> I am uncertain how we would qualify what it means to have complete information.

If CMake signals an "unknown" assembler (or compiler) during configuration, I would consider this an incomplete toolchain configuration (and the problem is in the _toolchain file_ if you specify one). Everything you do after this to fix issues is in the wrong place. From my point of view, a configuration that completes with an "unknown" assembler is a failed configuration because CMake cannot define how to handle the assembly source. Fixing this in the build script binds the build script to that specific (incomplete) environment specification. 

So, the first step in fixing the build should be to allow CMake to find the complete toolchain during the configuration.

> I am aware of this, and am mildly disappointed that VCPKG does not handle that when configured with VCPKG_TARGET_ARCHITECTURE=arm64. Not sure if I should be filing a bug upstream, what do you think?

Tbh, I cannot say what the expectation of `vcpkg` is in such a setup or why you use it, so I am not qualified to make the call whether this is a bug. But it certainly would help if they could verify that your configuration makes sense. 

If I were in your position, I would try to build sentry-native first without any of the `vcpkg` environmental setup and also without running `vcvarsall.bat` because the CMake generator will work very hard to find all the required tools for your configuration if you give it enough context (even without establishing a VS shell).

Enough context should be something like this:

https://github.com/getsentry/crashpad/blob/getsentry/.github/workflows/build.yml#L27-L47

The corresponding ARM64 configuration output:

https://github.com/getsentry/crashpad/actions/runs/7115516069/job/19371852568#step:6:21

And x64:

https://github.com/getsentry/crashpad/actions/runs/7115516069/job/19371852568#step:4:20

Just so you know, this is only our crashpad build. Still, besides the available configuration variables, it should work for `sentry-native` similarly. I cannot point you to a similar invocation in the `sentry-native` repository since we fully parameterize our CMake builds from Python for testing because there are specific tests for some of the build parameters. So, building with those parameters is part of the integration tests.

Only after you get it to run this way would I go to the next step and add the `vcpkg` configuration (or maybe only diffs of that configuration) because asking the `vcpkg` devs pointed questions from this baseline is much easier.

> So here I may be using incorrect invocation of CMake. As you see above, I use only `-A` at configuration stage along with CMake variables, which are directed to VCPKG. The only parameter which appears during configuration _as well as_ build, is the build configuration `RelWithDebInfo`. One question though, in the referenced section of your YAML file, I can see no mention of `CMAKE_GENERATOR_PLATFORM`. Instead I see `/p:Platform=ARM64`. Are these equivalent?

I was maybe a bit unclear by what I meant here:

* `CMAKE_GENERATOR_PLATFORM` is defined in the [toolchain configuration](https://github.com/getsentry/crashpad/blob/89991e9910bc4c0893e45c8cfad0bdd31cc25a5c/cmake/toolchains/win_arm64.cmake) which is passed to the CMake configuration phase.
* I prefer only to define `CMAKE_GENERATOR_PLATFORM` and leave out `-A` because that opens the door to conflicting parameters (if you change one, you'll have to change the other, not necessarily an issue in this particular build-script, but as I need to point users to these files as well, I want to reduce the amount of potential conflict).
* The problem is that while `CMAKE_GENERATOR_PLATFORM` is the correct configuration variable for what `-A` does when using the `MSVC` generator, it is only a configuration variable and does not affect the build.
* If you fully want to replace `-A`, then you also need to specify `/p:Platform=ARM64` for the build, or otherwise, it would build the `ARM64`-configured project as an `x64` project. :-D If you think this sounds ridiculous, then I can assure you that I fully agree.
* This means these two are not equivalent but support the different CMake phases, and together, they replace the `-A` parameter (if you need to get rid of it because you or `vcpkg` might be specifying a diverging `CMAKE_GENERATOR_PLATFORM` as part of the toolchain).

**Comment by deathlyrage** [comment]
I ran into this issue on the latest builds and commits. There is two problems.

CMAKE_SYSTEM_PROCESSOR returns AMD64 and CMAKE_GENERATOR_PLATFORM returns ARM64 when compiling from a AMD64 host. To fix this. I swap it out to use CMAKE_GENERATOR_PLATFORM.

If we apply the fix reccomended above to get around the bug in cmake. It will succesfully build.

[crashpad-windows-on-arm-fix.patch](https://github.com/user-attachments/files/17006586/crashpad-windows-on-arm-fix.patch)

This was tested using the latest cmake version 3.30.3.

Let me know i can open a new PR with this if it's needed.

**Comment by deathlyrage** [comment]
One more fix, CMAKE_SYSTEM_PROCESSOR is used in 2 more places that need to be changed.

```
 CMakeLists.txt | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index aff8a411..e86dc6d2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -65,8 +65,8 @@ function(crashpad_install_dev)
     endif()
 endfunction()
 
-if(WIN32 AND NOT (MINGW AND "${CMAKE_SYSTEM_PROCESSOR}" MATCHES "[Aa][Rr][Mm]64"))
-    if("${CMAKE_SYSTEM_PROCESSOR}" MATCHES ARM64)
+if(WIN32 AND NOT (MINGW AND "${CMAKE_GENERATOR_PLATFORM}" MATCHES "[Aa][Rr][Mm]64"))
+    if("${CMAKE_GENERATOR_PLATFORM}" MATCHES ARM64)
         enable_language(ASM_MARMASM)
     else()
         enable_language(ASM_MASM)
```

**Comment by past-due** [comment]
`CMAKE_GENERATOR_PLATFORM` will only work with Visual Studio (and only if the generator platform is specified with `cmake -A` or the `CMAKE_GENERATOR_PLATFORM` environment variable). So, among other issues, swapping this out will break MINGW builds.

I assume you are using Visual Studio? How exactly are you trying to compile the codebase? (Visual Studio Command Line prompt? Opening a CMake project in Visual Studio?)

**Comment by supervacuus** [comment]
@deathlyrage, sadly, the proposed fix—which is very similar to one from the OP—breaks various build configs. I tried to explain in detail in my above replies why it breaks behavior, why it assumes something that only a subset of generators provides, and why it indicates a toolchain configuration issue.

Given that you do not use `VCPKG,` could you create another issue where you highlight your cmake config and build invocations and show the resulting configuration logs?

But before you do that, can you give what I described [here](https://github.com/getsentry/sentry-native/issues/939#issuecomment-1919450693) another try?

If your configuration logs show a missing assembler, then the toolchain is incorrectly configured. You can verify this by checking the following output in your config logs:

```
-- The ASM_MASM compiler identification is unknown
-- Found assembler: ml
```

The correct way to cross-compile the Native SDK for ARM64 using the `msbuild/vs` generator is the following:

```
# specify a correctly configured toolchain, which we currently provide via crashpad (primarily for CI build)
cmake -B cmake-build-arm64 -DCMAKE_TOOLCHAIN_FILE="external/crashpad/cmake/toolchains/win_arm64.cmake"
# specify ARM64 explicitly as a passthrough parameter during the build (since replacing `-A` with `CMAKE_GENERATOR_PLATFORM` only affects the config invocation)
cmake --build cmake-build-arm64 --config RelWithDebInfo -- /p:Platform=ARM64
```

The referenced toolchain file contains the following (which addresses the order of initialization issue inherent to the msbuild generator when simply providing `-A`).
```
SET(CMAKE_GENERATOR_PLATFORM "ARM64")
SET(CMAKE_SYSTEM_PROCESSOR "ARM64")
SET(CMAKE_SYSTEM_NAME "Windows")
```

Which should lead to the correct assembler identification:

```
-- The ASM_MARMASM compiler identification is MSVC
-- Found assembler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/arm64/armasm64.exe
```

All issues I have seen reported either fail to set up the toolchain correctly or miss that the build invocation by default targets AMD64 (both problems inherent in the CMake + MSBuild combo).

I am happy to accept changes to the build script that ease the preconditions for a working config/build. I am reluctant to introduce a "fix" that alleviates a failed configuration.

**Comment by supervacuus** [comment]
> The correct way to cross-compile the Native SDK for ARM64 using the msbuild/vs generator is the following:
> ```
> # specify a correctly configured toolchain, which we currently provide via crashpad (primarily for CI build)
> cmake -B cmake-build-arm64 -DCMAKE_TOOLCHAIN_FILE="external/crashpad/cmake/toolchains/win_arm64.cmake"
> # specify ARM64 explicitly as a passthrough parameter during the build (since replacing `-A` with `CMAKE_GENERATOR_PLATFORM` only affects the config invocation)
> cmake --build cmake-build-arm64 --config RelWithDebInfo -- /p:Platform=ARM64
> ```

Btw, to be clear, you can still use `-A` here, but you cannot bypass setting `CMAKE_SYSTEM_PROCESSOR` via a toolchain file  (passing it via command line, i.e., as a cache-var is not sufficient!!!). I only remove `-A` to prevent setting  `CMAKE_GENERATOR_PLATFORM` in two - possibly diverging - places. The following is also correct:

```
# Set the MSBuild architecture (which also sets `CMAKE_GENERATOR_PLATFORM`), but specify `CMAKE_SYSTEM_PROCESSOR` via explicit toolchain
cmake -B cmake-build-arm64 -A ARM64 -DCMAKE_TOOLCHAIN_FILE="external/crashpad/cmake/toolchains/win_arm64.cmake"
# Setting `-A` in the config will automatically set the platform for the build so that we can ignore the pass-through
cmake --build cmake-build-arm64 --config RelWithDebInfo
```

Another aspect: if you use CMake with the Ninja generator, you can use `external/crashpad/cmake/toolchains/mingw_arm64.cmake` which not only works with MinGW, but all Ninja builds targeting ARM64 on Windows. We can merge the two CMake files, remove `CMAKE_GENERATOR_PLATFORM` from the toolchain and document specifying `-A` when using MSBuild if that is enough. 

However, I would need positive feedback from affected users before streamlining this and putting it into the docs. Unfortunately, I never got a response from the OP, either.
